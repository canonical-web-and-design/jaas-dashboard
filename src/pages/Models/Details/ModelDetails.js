import React, { useEffect, useMemo, useState } from "react";
import MainTable from "@canonical/react-components/dist/components/MainTable";
import { useDispatch, useSelector } from "react-redux";
import { useParams } from "react-router-dom";
import cloneDeep from "clone-deep";

import Filter from "components/Filter/Filter";
import InfoPanel from "components/InfoPanel/InfoPanel";
import Layout from "components/Layout/Layout";
import Terminal from "components/Terminal/Terminal";
import Header from "components/Header/Header";

import { getModelUUID, getModelStatus } from "app/selectors";
import { fetchModelStatus } from "juju/actions";
import { collapsibleSidebar } from "app/actions";
import {
  applicationTableHeaders,
  unitTableHeaders,
  machineTableHeaders,
  relationTableHeaders,
  generateApplicationRows,
  generateMachineRows,
  generateRelationRows,
  generateUnitRows
} from "./generators";

import "./_model-details.scss";

/**
  Returns the modelStatusData filtered by the supplied values.
  @param {Object} modelStatusData The model status data to filter
  @param {String} appName The name of the application to filter the data by.
*/
const filterModelStatusData = (modelStatusData, appName) => {
  if (!modelStatusData || appName === "") {
    return modelStatusData;
  }
  const filteredData = cloneDeep(modelStatusData);
  const application = filteredData.applications[appName];
  // remove the units from the application objects that are not
  // the filter-by app.
  const subordinateTo = application.subordinateTo || [];
  Object.keys(filteredData.applications).forEach(key => {
    if (key !== appName && !subordinateTo.includes(key)) {
      filteredData.applications[key].units = {};
    }
  });
  // Loop through all of the units of the remaining applications that are
  // listed in the subordinateTo list. This is done because although
  // a subordinate is supposed to be installed on each unit, that's not
  // always the case.
  subordinateTo.forEach(parentName => {
    const units = filteredData.applications[parentName].units;
    Object.entries(units).forEach(entry => {
      const found = Object.entries(entry[1].subordinates).find(
        ele => ele[0].split("/")[0] === appName
      );
      if (!found) {
        delete units[entry[0]];
      }
    });
  });

  // Remove all the machines that the selected application isn't installed on.
  const appMachines = new Set();
  for (let unitId in application.units) {
    const unit = application.units[unitId];
    appMachines.add(unit.machine);
  }
  subordinateTo.forEach(subAppName => {
    // this will be the parent of the subordinate and grab the machines from it
    const parent = filteredData.applications[subAppName];
    for (let unitId in parent.units) {
      const unit = parent.units[unitId];
      appMachines.add(unit.machine);
    }
  });
  for (let machineId in filteredData.machines) {
    if (!appMachines.has(machineId)) delete filteredData.machines[machineId];
  }

  // Remove all relations that don't involve the selected application.
  filteredData.relations = modelStatusData.relations.filter(
    relation => relation.key.indexOf(appName) > -1
  );

  return filteredData;
};

const ModelDetails = () => {
  const { 0: modelName } = useParams();
  const dispatch = useDispatch();
  const [filterByApp, setFilterByApp] = useState("");

  const getModelUUIDMemo = useMemo(() => getModelUUID(modelName), [modelName]);
  const modelUUID = useSelector(getModelUUIDMemo);
  const getModelStatusMemo = useMemo(() => getModelStatus(modelUUID), [
    modelUUID
  ]);
  const modelStatusData = useSelector(getModelStatusMemo);
  const filteredModelStatusData = filterModelStatusData(
    modelStatusData,
    filterByApp
  );

  useEffect(() => {
    dispatch(collapsibleSidebar(true));
    return () => {
      dispatch(collapsibleSidebar(false));
    };
  }, [dispatch]);

  useEffect(() => {
    if (modelUUID !== null && filteredModelStatusData === null) {
      // This model may not be in the first batch of models that we request
      // status from in the main loop so update the status now.
      dispatch(fetchModelStatus(modelUUID));
    }
  }, [dispatch, modelUUID, filteredModelStatusData]);

  const handleRowClick = e => {
    const currentTarget = e.currentTarget;
    if (currentTarget.classList.contains("is-selected")) {
      setFilterByApp("");
      return;
    }
    setFilterByApp(currentTarget.dataset.app);
  };

  const viewFilters = ["all", "apps", "units", "machines", "relations"];
  const statusFilters = ["all", "maintenance", "blocked"];

  const applicationTableRows = useMemo(
    () =>
      generateApplicationRows(
        filteredModelStatusData,
        filterByApp,
        handleRowClick
      ),
    [filterByApp, filteredModelStatusData]
  );
  const unitTableRows = useMemo(
    () => generateUnitRows(filteredModelStatusData, filterByApp),
    [filterByApp, filteredModelStatusData]
  );
  const machinesTableRows = useMemo(
    () => generateMachineRows(filteredModelStatusData, filterByApp),
    [filterByApp, filteredModelStatusData]
  );
  const relationTableRows = useMemo(
    () => generateRelationRows(filteredModelStatusData, filterByApp),
    [filterByApp, filteredModelStatusData]
  );

  return (
    <Layout>
      <Header>
        <div className="model-details__header">
          <strong>{modelStatusData ? modelStatusData.model.name : ""}</strong>
          <div className="model-details__filters">
            <Filter label="View:" filters={viewFilters} />
            <Filter label="Status:" filters={statusFilters} />
          </div>
        </div>
      </Header>
      <div className="l-content">
        <div className="model-details">
          <InfoPanel />
          <div className="model-details__main">
            <MainTable
              headers={applicationTableHeaders}
              rows={applicationTableRows}
              className="model-details__apps"
              sortable
            />
            <MainTable
              headers={unitTableHeaders}
              rows={unitTableRows}
              className="model-details__units"
              sortable
            />
            <MainTable
              headers={machineTableHeaders}
              rows={machinesTableRows}
              className="model-details__machines"
              sortable
            />
            <MainTable
              headers={relationTableHeaders}
              rows={relationTableRows}
              className="model-details__relations"
              sortable
            />
          </div>
        </div>
      </div>
      <Terminal
        address="wss://shell.jujugui.org:443/ws/"
        modelName={modelName}
      />
    </Layout>
  );
};

export default ModelDetails;
