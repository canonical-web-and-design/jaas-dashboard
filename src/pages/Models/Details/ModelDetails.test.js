import React from "react";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { mount } from "enzyme";
import { MemoryRouter } from "react-router";
import TestRoute from "components/Routes/TestRoute";
import dataDump from "testing/complete-redux-store-dump";

import ModelDetails from "./ModelDetails";

jest.mock("components/Terminal/Terminal", () => {
  const Terminal = () => <div className="terminal"></div>;
  return Terminal;
});

jest.mock("components/Topology/Topology", () => {
  const Topology = () => <div className="topology"></div>;
  return Topology;
});

const mockStore = configureStore([]);

describe("ModelDetail Container", () => {
  it("renders the details pane", () => {
    const store = mockStore(dataDump);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={["/models/group-test"]}>
          <TestRoute path="/models/*">
            <ModelDetails />
          </TestRoute>
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find("Topology").length).toBe(1);
    expect(wrapper.find("MainTable").length).toBe(4);
  });

  it("renders the details pane for models shared-with-me", () => {
    const store = mockStore(dataDump);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={["/models/space-man@external/frontend-ci"]}
        >
          <TestRoute path="/models/*">
            <ModelDetails />
          </TestRoute>
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find("MainTable").length).toBe(4);
  });

  it("renders the machine details section", () => {
    const store = mockStore(dataDump);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={["/models/spaceman@external/mymodel"]}>
          <TestRoute path="/models/*">
            <ModelDetails />
          </TestRoute>
        </MemoryRouter>
      </Provider>
    );
    expect(
      wrapper.find("MainTable").at(2).hasClass("model-details__machines")
    ).toBe(true);
  });

  it("subordinate rows render correct amount", () => {
    const store = mockStore(dataDump);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={["/models/sub-test"]}>
          <TestRoute path="/models/*">
            <ModelDetails />
          </TestRoute>
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find(".subordinate").length).toEqual(2);
  });

  it("renders the terminal", () => {
    const store = mockStore(dataDump);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={["/models/test1"]}>
          <TestRoute path="/models/*">
            <ModelDetails />
          </TestRoute>
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find("Terminal").length).toBe(1);
  });

  it("clicking an application row filters the results", () => {
    const store = mockStore(dataDump);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={["/models/activedev@external/sub-test"]}>
          <TestRoute path="/models/*">
            <ModelDetails />
          </TestRoute>
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find(".model-details__units tbody tr").length).toBe(4);
    wrapper
      .find('.model-details__apps tr[data-app="easyrsa"]')
      .simulate("click");
    const units = wrapper.find(".model-details__units tbody tr");
    expect(units.length).toBe(1);
    expect(units.hasClass("is-selected"));
    expect(units.find("td").first().text()).toBe("easyrsa/0");
    const machines = wrapper.find(".model-details__machines tbody tr");
    expect(machines.length).toBe(1);
    expect(machines.find("td").first().text()).toBe("135.229.83.62");
    expect(wrapper.find(".model-details__relations tbody tr").length).toBe(0);
  });

  it("clicking an application row filters the results (subordinates)", () => {
    const store = mockStore(dataDump);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={["/models/activedev@external/sub-test"]}>
          <TestRoute path="/models/*">
            <ModelDetails />
          </TestRoute>
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find(".model-details__units tbody tr").length).toBe(4);
    wrapper
      .find('.model-details__apps tr[data-app="telegraf"]')
      .simulate("click");
    const units = wrapper.find(".model-details__units tbody tr");
    expect(units.length).toBe(3);
    expect(units.first().find("td").at(0).text()).toBe("ubuntu/0");
    expect(units.at(1).find("td").first().text()).toBe("nrpe/0");
    expect(units.at(2).find("td").first().text()).toBe("telegraf/0");
    const machines = wrapper.find(".model-details__machines tbody tr");
    expect(machines.length).toBe(1);
    expect(machines.find("td").first().text()).toBe("035.243.128.238");
    const relations = wrapper.find(".model-details__relations tbody tr");
    expect(relations.length).toBe(1);
    expect(relations.find("td").first().text()).toBe("ubuntu:juju-info");
  });

  it("clicking an application row filters the results (missing subordinates)", () => {
    // This is for when an application has subordinates related to it but not
    // all units have those subordinates for whatever reason.
    const store = mockStore(dataDump);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={["/models/spaceman@external/hadoopspark"]}
        >
          <TestRoute path="/models/*">
            <ModelDetails />
          </TestRoute>
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find(".model-details__units tbody tr").length).toBe(12);
    wrapper
      .find('.model-details__apps tr[data-app="rsyslog-forwarder-ha"]')
      .simulate("click");
    const units = wrapper.find(".model-details__units tbody tr");
    expect(units.length).toBe(3);
    expect(units.first().find("td").at(0).text()).toBe("slave/1");
    expect(units.at(1).find("td").first().text()).toBe("ganglia-node/0");
    expect(units.at(2).find("td").first().text()).toBe(
      "rsyslog-forwarder-ha/0"
    );
    const machines = wrapper.find(".model-details__machines tbody tr");
    expect(machines.length).toBe(1);
    expect(machines.find("td").first().text()).toBe("035.227.34.90");
    const relations = wrapper.find(".model-details__relations tbody tr");
    expect(relations.length).toBe(4);
    expect(relations.first().find("td").first().text()).toBe(
      "rsyslog:aggregator"
    );
    expect(relations.at(1).find("td").first().text()).toBe(
      "namenode:juju-info"
    );
    expect(relations.at(2).find("td").first().text()).toBe("slave:juju-info");
    expect(relations.at(3).find("td").first().text()).toBe(
      "resourcemanager:juju-info"
    );
  });

  it("view filters hide and show tables", () => {
    const store = mockStore(dataDump);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={["/models/spaceman@external/hadoopspark"]}
        >
          <TestRoute path="/models/*">
            <ModelDetails />
          </TestRoute>
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find("MainTable").length).toBe(4);
    wrapper.find("[data-test='apps'] button").simulate("click");
    expect(wrapper.find("MainTable").length).toBe(1);
    expect(wrapper.find("table.model-details__apps").length).toBe(1);
    wrapper.find("[data-test='apps'] button").simulate("click");
    expect(wrapper.find("MainTable").length).toBe(4);
    wrapper.find("[data-test='machines'] button").simulate("click");
    expect(wrapper.find("table.model-details__machines").length).toBe(1);
    wrapper.find("[data-test='relations'] button").simulate("click");
    expect(wrapper.find("table.model-details__relations").length).toBe(1);
    expect(wrapper.find("MainTable").length).toBe(2);
  });

  it("app names link to charm store pages", () => {
    const store = mockStore(dataDump);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={["/models/spaceman@external/hadoopspark"]}
        >
          <TestRoute path="/models/*">
            <ModelDetails />
          </TestRoute>
        </MemoryRouter>
      </Provider>
    );
    const appsFirstRowLink = wrapper
      .find(".model-details__apps tr [data-test='app-link']")
      .at(1);
    expect(appsFirstRowLink.prop("href")).toEqual(
      "https://www.jaas.ai/u/activedev/failtester/precise/7"
    );
  });
});
