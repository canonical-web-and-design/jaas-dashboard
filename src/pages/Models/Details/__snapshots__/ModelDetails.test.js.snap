// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelDetail Container renders the details pane when the user is logged in 1`] = `
<div
  className="model-details"
>
  <InfoPanel>
    <div
      className="info-panel"
    >
      <div
        className="info-panel__grid"
      >
        <div
          className="info-panel__grid-item"
          key="Model"
        >
          <h4
            className="p-muted-heading"
          >
            Model
          </h4>
          <p>
            cdk-default
          </p>
        </div>
        <div
          className="info-panel__grid-item"
          key="Region"
        >
          <h4
            className="p-muted-heading"
          >
            Region
          </h4>
          <p>
            us-east-1
          </p>
        </div>
        <div
          className="info-panel__grid-item"
          key="Controller"
        >
          <h4
            className="p-muted-heading"
          >
            Controller
          </h4>
          <p>
            aws-controller
          </p>
        </div>
        <div
          className="info-panel__grid-item"
          key="Cloud"
        >
          <h4
            className="p-muted-heading"
          >
            Cloud
          </h4>
          <p>
            aws
          </p>
        </div>
        <div
          className="info-panel__grid-item"
          key="SLA"
        >
          <h4
            className="p-muted-heading"
          >
            SLA
          </h4>
          <p>
            Unsupported
          </p>
        </div>
      </div>
      <img
        alt=""
        src="https://assets.ubuntu.com/v1/f3f75945-home-promo-bundle.svg"
      />
    </div>
  </InfoPanel>
  <div
    className="model-details__main"
  >
    <div
      className="model-details__filters"
    >
      <Filter
        filters={
          Array [
            "all",
            "apps",
            "units",
            "machines",
            "relations",
          ]
        }
        label="View"
      >
        <div
          className="p-filter"
        >
          <span
            className="p-filter__label"
          >
            View
          </span>
          <span
            className="p-filter__item"
            key="all"
          >
            <Button>
              <button
                className="p-button--filter"
              >
                all
              </button>
            </Button>
          </span>
          <span
            className="p-filter__item"
            key="apps"
          >
            <Button>
              <button
                className="p-button--filter"
              >
                apps
              </button>
            </Button>
          </span>
          <span
            className="p-filter__item"
            key="units"
          >
            <Button>
              <button
                className="p-button--filter"
              >
                units
              </button>
            </Button>
          </span>
          <span
            className="p-filter__item"
            key="machines"
          >
            <Button>
              <button
                className="p-button--filter"
              >
                machines
              </button>
            </Button>
          </span>
          <span
            className="p-filter__item"
            key="relations"
          >
            <Button>
              <button
                className="p-button--filter"
              >
                relations
              </button>
            </Button>
          </span>
        </div>
      </Filter>
      <Filter
        filters={
          Array [
            "all",
            "maintenance",
            "blocked",
          ]
        }
        label="Status"
      >
        <div
          className="p-filter"
        >
          <span
            className="p-filter__label"
          >
            Status
          </span>
          <span
            className="p-filter__item"
            key="all"
          >
            <Button>
              <button
                className="p-button--filter"
              >
                all
              </button>
            </Button>
          </span>
          <span
            className="p-filter__item"
            key="maintenance"
          >
            <Button>
              <button
                className="p-button--filter"
              >
                maintenance
              </button>
            </Button>
          </span>
          <span
            className="p-filter__item"
            key="blocked"
          >
            <Button>
              <button
                className="p-button--filter"
              >
                blocked
              </button>
            </Button>
          </span>
        </div>
      </Filter>
    </div>
    <MainTable
      headers={
        Array [
          Object {
            "content": "App",
            "sortKey": "app",
          },
          Object {
            "content": "Status",
            "sortKey": "status",
          },
          Object {
            "content": "Version",
            "sortKey": "version",
          },
          Object {
            "className": "u-align--right",
            "content": "Scale",
            "sortKey": "scale",
          },
          Object {
            "content": "Store",
            "sortKey": "store",
          },
          Object {
            "className": "u-align--right",
            "content": "Rev",
            "sortKey": "rev",
          },
          Object {
            "content": "OS",
            "sortKey": "os",
          },
          Object {
            "content": "Notes",
            "sortKey": "notes",
          },
        ]
      }
      rows={
        Array [
          Object {
            "columns": Array [
              Object {
                "content": "Ready",
                "role": "rowheader",
              },
              Object {
                "content": 1,
              },
              Object {
                "content": "1 GiB",
              },
              Object {
                "className": "u-align--right",
                "content": 2,
              },
              Object {
                "content": "Ready",
              },
              Object {
                "className": "u-align--right",
                "content": 1,
              },
              Object {
                "content": "1 GiB",
              },
              Object {
                "content": 2,
              },
            ],
            "sortData": Object {
              "app": "ready",
              "notes": 2,
              "os": 1,
              "rev": 2,
              "scale": 2,
              "status": 2,
              "store": "ready",
              "version": 1,
            },
          },
          Object {
            "columns": Array [
              Object {
                "content": "Idle",
                "role": "rowheader",
              },
              Object {
                "className": "u-align--right",
                "content": 1,
              },
              Object {
                "className": "u-align--right",
                "content": "1 GiB",
              },
              Object {
                "className": "u-align--right",
                "content": 2,
              },
              Object {
                "content": "Ready",
              },
              Object {
                "className": "u-align--right",
                "content": 1,
              },
              Object {
                "content": "1 GiB",
              },
              Object {
                "content": 2,
              },
            ],
            "sortData": Object {
              "cores": 1,
              "disks": 2,
              "ram": 1,
              "status": "idle",
            },
          },
          Object {
            "columns": Array [
              Object {
                "content": "Waiting",
                "role": "rowheader",
              },
              Object {
                "content": 8,
              },
              Object {
                "content": "3.9 GiB",
              },
              Object {
                "className": "u-align--right",
                "content": 3,
              },
              Object {
                "content": "Ready",
              },
              Object {
                "className": "u-align--right",
                "content": 1,
              },
              Object {
                "content": "1 GiB",
              },
              Object {
                "content": 2,
              },
            ],
            "sortData": Object {
              "cores": 8,
              "disks": 3,
              "ram": 3.9,
              "status": "waiting",
            },
          },
        ]
      }
      sortable={true}
    >
      <Table
        sortable={true}
      >
        <table
          className="p-table--sortable"
          role="grid"
        >
          <thead>
            <TableRow>
              <tr
                role="row"
              >
                <TableHeader
                  key="0"
                  onClick={[Function]}
                  sort="none"
                >
                  <th
                    aria-sort="none"
                    onClick={[Function]}
                    role="columnheader"
                  >
                    App
                  </th>
                </TableHeader>
                <TableHeader
                  key="1"
                  onClick={[Function]}
                  sort="none"
                >
                  <th
                    aria-sort="none"
                    onClick={[Function]}
                    role="columnheader"
                  >
                    Status
                  </th>
                </TableHeader>
                <TableHeader
                  key="2"
                  onClick={[Function]}
                  sort="none"
                >
                  <th
                    aria-sort="none"
                    onClick={[Function]}
                    role="columnheader"
                  >
                    Version
                  </th>
                </TableHeader>
                <TableHeader
                  className="u-align--right"
                  key="3"
                  onClick={[Function]}
                  sort="none"
                >
                  <th
                    aria-sort="none"
                    className="u-align--right"
                    onClick={[Function]}
                    role="columnheader"
                  >
                    Scale
                  </th>
                </TableHeader>
                <TableHeader
                  key="4"
                  onClick={[Function]}
                  sort="none"
                >
                  <th
                    aria-sort="none"
                    onClick={[Function]}
                    role="columnheader"
                  >
                    Store
                  </th>
                </TableHeader>
                <TableHeader
                  className="u-align--right"
                  key="5"
                  onClick={[Function]}
                  sort="none"
                >
                  <th
                    aria-sort="none"
                    className="u-align--right"
                    onClick={[Function]}
                    role="columnheader"
                  >
                    Rev
                  </th>
                </TableHeader>
                <TableHeader
                  key="6"
                  onClick={[Function]}
                  sort="none"
                >
                  <th
                    aria-sort="none"
                    onClick={[Function]}
                    role="columnheader"
                  >
                    OS
                  </th>
                </TableHeader>
                <TableHeader
                  key="7"
                  onClick={[Function]}
                  sort="none"
                >
                  <th
                    aria-sort="none"
                    onClick={[Function]}
                    role="columnheader"
                  >
                    Notes
                  </th>
                </TableHeader>
              </tr>
            </TableRow>
          </thead>
          <tbody>
            <TableRow
              key="0"
            >
              <tr
                role="row"
              >
                <TableCell
                  key="0"
                  role="rowheader"
                >
                  <td
                    className=""
                    role="rowheader"
                  >
                    Ready
                  </td>
                </TableCell>
                <TableCell
                  key="1"
                >
                  <td
                    className=""
                    role="gridcell"
                  >
                    1
                  </td>
                </TableCell>
                <TableCell
                  key="2"
                >
                  <td
                    className=""
                    role="gridcell"
                  >
                    1 GiB
                  </td>
                </TableCell>
                <TableCell
                  className="u-align--right"
                  key="3"
                >
                  <td
                    className="u-align--right"
                    role="gridcell"
                  >
                    2
                  </td>
                </TableCell>
                <TableCell
                  key="4"
                >
                  <td
                    className=""
                    role="gridcell"
                  >
                    Ready
                  </td>
                </TableCell>
                <TableCell
                  className="u-align--right"
                  key="5"
                >
                  <td
                    className="u-align--right"
                    role="gridcell"
                  >
                    1
                  </td>
                </TableCell>
                <TableCell
                  key="6"
                >
                  <td
                    className=""
                    role="gridcell"
                  >
                    1 GiB
                  </td>
                </TableCell>
                <TableCell
                  key="7"
                >
                  <td
                    className=""
                    role="gridcell"
                  >
                    2
                  </td>
                </TableCell>
              </tr>
            </TableRow>
            <TableRow
              key="1"
            >
              <tr
                role="row"
              >
                <TableCell
                  key="0"
                  role="rowheader"
                >
                  <td
                    className=""
                    role="rowheader"
                  >
                    Idle
                  </td>
                </TableCell>
                <TableCell
                  className="u-align--right"
                  key="1"
                >
                  <td
                    className="u-align--right"
                    role="gridcell"
                  >
                    1
                  </td>
                </TableCell>
                <TableCell
                  className="u-align--right"
                  key="2"
                >
                  <td
                    className="u-align--right"
                    role="gridcell"
                  >
                    1 GiB
                  </td>
                </TableCell>
                <TableCell
                  className="u-align--right"
                  key="3"
                >
                  <td
                    className="u-align--right"
                    role="gridcell"
                  >
                    2
                  </td>
                </TableCell>
                <TableCell
                  key="4"
                >
                  <td
                    className=""
                    role="gridcell"
                  >
                    Ready
                  </td>
                </TableCell>
                <TableCell
                  className="u-align--right"
                  key="5"
                >
                  <td
                    className="u-align--right"
                    role="gridcell"
                  >
                    1
                  </td>
                </TableCell>
                <TableCell
                  key="6"
                >
                  <td
                    className=""
                    role="gridcell"
                  >
                    1 GiB
                  </td>
                </TableCell>
                <TableCell
                  key="7"
                >
                  <td
                    className=""
                    role="gridcell"
                  >
                    2
                  </td>
                </TableCell>
              </tr>
            </TableRow>
            <TableRow
              key="2"
            >
              <tr
                role="row"
              >
                <TableCell
                  key="0"
                  role="rowheader"
                >
                  <td
                    className=""
                    role="rowheader"
                  >
                    Waiting
                  </td>
                </TableCell>
                <TableCell
                  key="1"
                >
                  <td
                    className=""
                    role="gridcell"
                  >
                    8
                  </td>
                </TableCell>
                <TableCell
                  key="2"
                >
                  <td
                    className=""
                    role="gridcell"
                  >
                    3.9 GiB
                  </td>
                </TableCell>
                <TableCell
                  className="u-align--right"
                  key="3"
                >
                  <td
                    className="u-align--right"
                    role="gridcell"
                  >
                    3
                  </td>
                </TableCell>
                <TableCell
                  key="4"
                >
                  <td
                    className=""
                    role="gridcell"
                  >
                    Ready
                  </td>
                </TableCell>
                <TableCell
                  className="u-align--right"
                  key="5"
                >
                  <td
                    className="u-align--right"
                    role="gridcell"
                  >
                    1
                  </td>
                </TableCell>
                <TableCell
                  key="6"
                >
                  <td
                    className=""
                    role="gridcell"
                  >
                    1 GiB
                  </td>
                </TableCell>
                <TableCell
                  key="7"
                >
                  <td
                    className=""
                    role="gridcell"
                  >
                    2
                  </td>
                </TableCell>
              </tr>
            </TableRow>
          </tbody>
        </table>
      </Table>
    </MainTable>
  </div>
</div>
`;

exports[`ModelDetail Container renders with a login message when not logged in 1`] = `
Array [
  <div
    className="login"
  >
    <h1
      className="login__heading"
    >
      <div
        className="p-media-object"
      >
        <img
          alt=""
          className="p-media-object__image"
          src="https://assets.ubuntu.com/v1/60d9b81e-picto-canonical.svg"
          width="50"
        />
        <div
          className="p-media-object__details"
        >
          <h3
            className="p-media-object__title"
          >
            Canonical Candid
          </h3>
        </div>
      </div>
    </h1>
    <div
      className="p-card--highlighted"
    >
      <h3>
        Login
      </h3>
      <div
        className="p-media-object"
      >
        <img
          alt=""
          className="p-media-object__image"
          src="https://assets.ubuntu.com/v1/c4f35e06-products-hero-ubuntu.svg"
          width="50"
        />
        <div
          className="p-media-object__details"
        >
          <h3
            className="p-media-object__title"
          >
            Ubuntu SSO
          </h3>
        </div>
      </div>
      <p>
        Will authenticate with Candid in order to log you in.
      </p>
      <button
        className="p-button--positive"
      >
        Proceed
      </button>
    </div>
  </div>,
  <div
    className="p-media-object"
  >
    <img
      alt=""
      className="p-media-object__image"
      src="https://assets.ubuntu.com/v1/60d9b81e-picto-canonical.svg"
      width="50"
    />
    <div
      className="p-media-object__details"
    >
      <h3
        className="p-media-object__title"
      >
        Canonical Candid
      </h3>
    </div>
  </div>,
  <div
    className="p-media-object__details"
  >
    <h3
      className="p-media-object__title"
    >
      Canonical Candid
    </h3>
  </div>,
  <div
    className="p-card--highlighted"
  >
    <h3>
      Login
    </h3>
    <div
      className="p-media-object"
    >
      <img
        alt=""
        className="p-media-object__image"
        src="https://assets.ubuntu.com/v1/c4f35e06-products-hero-ubuntu.svg"
        width="50"
      />
      <div
        className="p-media-object__details"
      >
        <h3
          className="p-media-object__title"
        >
          Ubuntu SSO
        </h3>
      </div>
    </div>
    <p>
      Will authenticate with Candid in order to log you in.
    </p>
    <button
      className="p-button--positive"
    >
      Proceed
    </button>
  </div>,
  <div
    className="p-media-object"
  >
    <img
      alt=""
      className="p-media-object__image"
      src="https://assets.ubuntu.com/v1/c4f35e06-products-hero-ubuntu.svg"
      width="50"
    />
    <div
      className="p-media-object__details"
    >
      <h3
        className="p-media-object__title"
      >
        Ubuntu SSO
      </h3>
    </div>
  </div>,
  <div
    className="p-media-object__details"
  >
    <h3
      className="p-media-object__title"
    >
      Ubuntu SSO
    </h3>
  </div>,
]
`;
